core/modfunc.go

```diff
@@ -389,18 +391,34 @@ func (fn *ModuleFunction) Call(ctx context.Context, opts *CallOpts) (t dagql.Typ
 		return nil, fmt.Errorf("failed to collect IDs: %w", err)
 	}
 
-	for _, id := range returnedIDs {
-		if err := fn.root.AddClientResourcesFromID(ctx, id, clientID, false); err != nil {
-			return nil, fmt.Errorf("failed to add client resources from ID: %w", err)
-		}
-	}
-
 	// NOTE: once generalized function caching is enabled we need to ensure that any non-reproducible
 	// cache entries are linked to the result of this call.
 	// See the previous implementation of this for a reference:
 	// https://github.com/dagger/dagger/blob/7c31db76e07c9a17fcdb3f3c4513c915344c1da8/core/modfunc.go#L483
 
-	return returnValueTyped, nil
+	// Function calls are cached per-session, but every client caller needs to add
+	// secret/socket/etc. resources from the result to their store.
+	callerClientMemo := sync.Map{}
+	return &dagql.PostCallTyped{
+		Typed: returnValueTyped,
+		PostCall: func(ctx context.Context) error {
+			// only run this once per calling client, no need to re-add resources
+			clientMetadata, err := engine.ClientMetadataFromContext(ctx)
+			if err != nil {
+				return fmt.Errorf("failed to get client metadata: %w", err)
+			}
+			if _, alreadyRan := callerClientMemo.LoadOrStore(clientMetadata.ClientID, struct{}{}); alreadyRan {
+				return nil
+			}
+
+			for _, id := range returnedIDs {
+				if err := fn.root.AddClientResourcesFromID(ctx, id, clientID, false); err != nil {
+					return fmt.Errorf("failed to add client resources from ID: %w", err)
+				}
```

Comment:
question: does this also affect the changes in https://github.com/dagger/dagger/pull/9697?

For example, can `LoadContext` also end up being cached in some cases where we'd need to do a similar transfer?

Reply:
Good question but as far as I can think, it shouldn't matter. The changes in this PR ensure that the caller always loads secrets in *return* values whether or not the call itself was cached. So then the scenarios are:
1. Call is not cached: the secrets needed for the contextual args will be loaded since it's gonna execute
2. Call *is* cached: contextual args aren't needed at all since the call is cached. If any of the secrets referenced by the contextual args end up referenced in the return value, then the secrets will be transferred.